src/libos/src/vm/vm_manager.rs

```diff
@@ -61,16 +75,62 @@ impl VMManager {
     }
 
     // Allocate single VMA chunk for new process whose process VM is not ready yet
+    // block_on is used in this function because making this function async will ultimately cause the ecall
+    // to be async which we don't support for now.
+    // TODO: Make this function Async.
     pub fn alloc(&self, options: &VMMapOptions) -> Result<(VMRange, ChunkRef)> {
-        let addr = *options.addr();
-        let size = *options.size();
-        if let Ok(new_chunk) = self.internal().mmap_chunk(options) {
-            return Ok((new_chunk.range().clone(), new_chunk));
+        let ret = self.try_alloc(options);
+        if !ret.has_errno(ENOMEM) || !vm_clean::has_potential_free_memory() {
+            return ret;
+        }
+
+        // Help clean and try again
+        vm_clean::help_dirty_range_clean();
+        warn!("alloc try second time");
+        let ret = self.try_alloc(options);
+        if !ret.has_errno(ENOMEM) {
+            return ret;
         }
-        return_errno!(ENOMEM, "can't allocate free chunks");
+
+        // In the second attempt, the current thread helps clean the dirty ranges. But didn't wait
+        // for all the ranges to be cleaned. In the third try, wait until there is no dirty range or
+        // ranges being cleaned. And then try again.
+        // It only reaches here when the memory resource is extremely constrained.
+        warn!("memory resource low, wait for clean memory if any and then try again");
+        async_rt::task::block_on(vm_clean::clean_target_dirty_range_if_any(
```

Comment:
block_on here could cause deadlock. Maybe move to ecall_new_process. Please make sure this will not cause a deadlock when there is only one CPU.

Reply:
Block on need a static lifetime which is troublesome to handle in the ecall. Remove block on. Only try two times here. Add more comments.